core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -160,13 +160,16 @@ inline void configure_shmem_preference(KernelFuncPtr const& func,
                                        bool prefer_shmem) {
 #ifndef KOKKOS_ARCH_KEPLER
   // On Kepler the L1 has no benefit since it doesn't cache reads
-  static bool cache_config_preference_cached = [&] {
+  auto set_cache_config() = [&] {
     CUDA_SAFE_CALL(cudaFuncSetCacheConfig(
         func,
         (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)));
     return prefer_shmem;
-  }();
-  (void)cache_config_preference_cached;
+  };
+  static bool cache_config_preference_cached = set_cache_config();
+  if (cache_config_preference_cached != prefer_shmem) {
+    cache_config_preference_cached = set_cache_config();
```

Comment:
We can't just do this because it's not thread-safe, which was the original reason why we're using an immediately evaluated lambda in the static initializer.

Reply:
I will just revert 3e6b6f7

Reply:
This doesn't change anything. The update is still not thread-safe, and now there's duplicated code.